config/initializers/assets.rb

```diff
@@ -1,105 +1,111 @@
+
+  # Serve plugin static assets if the application is configured to do so
+  if config.serve_static_assets
+    ::Rails::Engine.subclasses.map(&:instance).each do |engine|
+      if File.exist?("#{engine.root}/public/assets")
+        app.middleware.use ::ActionDispatch::Static, "#{engine.root}/public"
+      end
     end
   end
-end
 
-# Adds plugin assets to the application digests hash if a manifest file exists for a plugin
-config.after_initialize do
-  if (manifest_file = Dir.glob("#{Rails.root}/public/assets/manifest*.json").first)
-    foreman_manifest = JSON.parse(File.read(manifest_file))
+  # Adds plugin assets to the application digests hash if a manifest file exists for a plugin
+  config.after_initialize do
+    if (manifest_file = Dir.glob("#{Rails.root}/public/assets/manifest*.json").first)
+      foreman_manifest = JSON.parse(File.read(manifest_file))
 
-    ::Rails::Engine.subclasses.map(&:instance).each do |engine|
-      [engine.root, app.root].each do |root_dir|
-        manifest_path = File.join(root_dir, "public/assets/#{engine.engine_name}/#{engine.engine_name}.json")
+      ::Rails::Engine.subclasses.map(&:instance).each do |engine|
+        [engine.root, app.root].each do |root_dir|
+          manifest_path = File.join(root_dir, "public/assets/#{engine.engine_name}/#{engine.engine_name}.json")
 
-        if File.file?(manifest_path)
-          assets = JSON.parse(File.read(manifest_path))
+          if File.file?(manifest_path)
+            assets = JSON.parse(File.read(manifest_path))
 
-          foreman_manifest['files']  = foreman_manifest['files'].merge(assets['files'])
-          foreman_manifest['assets'] = foreman_manifest['assets'].merge(assets['assets'])
+            foreman_manifest['files']  = foreman_manifest['files'].merge(assets['files'])
+            foreman_manifest['assets'] = foreman_manifest['assets'].merge(assets['assets'])
+          end
         end
       end
-    end
 
-    manifest_filename = 'asset_manifest.json'
-    tmp_path          = "#{Rails.root}/tmp"
+      manifest_filename = 'asset_manifest.json'
+      tmp_path          = "#{Rails.root}/tmp"
 
-    Tempfile.open(manifest_filename, tmp_path) do |file|
-      file.write(foreman_manifest.to_json)
-      app.assets_manifest              = Sprockets::Manifest.new(app.assets, file.path)
-      ActionView::Base.assets_manifest = app.assets_manifest
+      Tempfile.open(manifest_filename, tmp_path) do |file|
+        file.write(foreman_manifest.to_json)
+        app.assets_manifest              = Sprockets::Manifest.new(app.assets, file.path)
+        ActionView::Base.assets_manifest = app.assets_manifest
+      end
     end
   end
+
+  # Append plugins asset paths to config.assets.precompile
+  asset_initializers = app.railties.map(&:initializers).flatten.
+    select do |initializer|
+    # This initializer puts all assets paths in
+    # SETTINGS[:plugin_name][:assets][:precompile]
+    initializer.name =~ /configure_assets/
+  end
```

Comment:
Hmm, I can't reproduce that.  The plugin example seems fine to me, I'm testing with a simple plugin that adds a single Javascript file to both SETTINGS and app.config.assets.precompile.

If I remove this bit of code that's adding to the precompile list in core and remove the precompile addition in the plugin, it correctly raises an exception when trying to use that JS.  If I restore the plugin's precompile addition (https://github.com/theforeman/foreman_plugin_template/blob/master/lib/foreman_plugin_template/engine.rb#L52-L54), it's added to the precompile list late, but doesn't raise an exception.

All plugins seem to have this with the exception perhaps of bastion.  I'm not sure I see why this is needed generally.